querybook/server/logic/datadoc_permission.py

```diff
@@ -13,34 +19,53 @@ class DocDoesNotExist(Exception):
 
 @with_session
 def user_can_write(doc_id, uid, session=None):
-    doc, editor = session.query(DataDoc, DataDocEditor).outerjoin(
-        DataDocEditor,
-        and_(DataDoc.id == DataDocEditor.data_doc_id, DataDocEditor.uid == uid),
-    ).filter(DataDoc.id == doc_id).first() or (None, None)
-
-    if doc is None:
-        raise DocDoesNotExist()
+    datadoc = session.query(DataDoc).filter_by(id=doc_id).first()
+    if datadoc.owner_uid == uid:
+        return True
 
-    if doc.owner_uid == uid:
+    editor = session.query(DataDocEditor).filter_by(uid=uid, data_doc_id=doc_id).first()
+    if editor is not None and editor.write:
         return True
 
-    return editor is not None and editor.write
+    inherited_editors = get_all_groups_and_group_members_with_access(
+        doc_or_board_id=doc_id,
+        editor_type=DataDocEditor,
+        uid=uid,
+        session=session,
+    )
+
+    if len(inherited_editors) == 1:
+        # Check if the editor's write privileges are true
+        if inherited_editors[0][3]:
+            return True
```

Comment:
the user permission checking logic, or even the whole function for datadoc and board seems quite similar, wondering if we can have something resuable